lib/datasource/maven/index.ts

```diff
@@ -177,162 +80,286 @@ function extractSnapshotVersion(metadata: XmlDocument): string | null {
   return `${version}-${timestamp}-${build}`;
 }
 
-async function getSnapshotFullVersion(
-  version: string,
-  dependency: MavenDependency,
-  repoUrl: string
-): Promise<string | null> {
-  // To determine what actual files are available for the snapshot, first we have to fetch and parse
-  // the metadata located at http://<repo>/<group>/<artifact>/<version-SNAPSHOT>/maven-metadata.xml
-  const metadataUrl = getMavenUrl(
-    dependency,
-    repoUrl,
-    `${version}/maven-metadata.xml`
-  );
+export const defaultRegistryUrls = [MAVEN_REPO];
 
-  const { xml: mavenMetadata } = await downloadMavenXml(metadataUrl);
-  if (!mavenMetadata) {
-    return null;
-  }
+export class MavenDatasource extends Datasource {
+  static id = 'maven';
 
-  return extractSnapshotVersion(mavenMetadata);
-}
+  override readonly defaultRegistryUrls = defaultRegistryUrls;
 
-async function createUrlForDependencyPom(
-  version: string,
-  dependency: MavenDependency,
-  repoUrl: string
-): Promise<string> {
-  if (isSnapshotVersion(version)) {
-    // By default, Maven snapshots are deployed to the repository with fixed file names.
-    // Resolve the full, actual pom file name for the version.
-    const fullVersion = await getSnapshotFullVersion(
-      version,
-      dependency,
-      repoUrl
-    );
+  override readonly defaultVersioning = mavenVersioning.id;
 
-    // If we were able to resolve the version, use that, otherwise fall back to using -SNAPSHOT
-    if (fullVersion !== null) {
-      return `${version}/${dependency.name}-${fullVersion}.pom`;
-    }
+  override readonly registryStrategy = 'merge';
+
+  constructor(id = MavenDatasource.id) {
+    super(id);
   }
 
-  return `${version}/${dependency.name}-${version}.pom`;
-}
+  async fetchReleasesFromMetadata(
+    dependency: MavenDependency,
+    repoUrl: string
+  ): Promise<ReleaseMap> {
+    const metadataUrl = getMavenUrl(dependency, repoUrl, 'maven-metadata.xml');
+
+    const cacheNamespace = 'datasource-maven:metadata-xml';
+    const cacheKey = metadataUrl.toString();
+    const cachedVersions = await packageCache.get<ReleaseMap>(
+      cacheNamespace,
+      cacheKey
+    );
+    /* istanbul ignore if */
+    if (cachedVersions) {
+      return cachedVersions;
+    }
 
-async function addReleasesUsingHeadRequests(
-  inputReleaseMap: ReleaseMap,
-  dependency: MavenDependency,
-  repoUrl: string
-): Promise<ReleaseMap> {
-  const releaseMap = { ...inputReleaseMap };
+    const { authorization, xml: mavenMetadata } = await downloadMavenXml(
+      this.http,
+      metadataUrl
+    );
+    if (!mavenMetadata) {
+      return {};
+    }
 
-  if (process.env.RENOVATE_EXPERIMENTAL_NO_MAVEN_POM_CHECK) {
+    const versions = extractVersions(mavenMetadata);
+    const releaseMap = versions.reduce(
+      (acc, version) => ({ ...acc, [version]: null }),
+      {}
+    );
+    if (!authorization) {
+      await packageCache.set(cacheNamespace, cacheKey, releaseMap, 30);
```

Comment:
This is why I didn't use decorator-based cache. Let's do this later maybe.

Reply:
Is it possible to use the `cacheable` property of the decorator?

https://github.com/renovatebot/renovate/blob/e7a55d2d37485eeb8d5320c131be572d2cdaed28/lib/util/cache/package/decorator.ts#L79-L83

It's used to only cache the default registry in `crate`

https://github.com/renovatebot/renovate/blob/c1555efb3b6e273ee6771f7b9ca746f5dad4eab2/lib/datasource/crate/index.ts#L35-L36

Reply:
I think it's possible, but decorator usage for Maven may require a pretty big refactoring, and I would like to do this as the separate step.

Reply:
Sounds good. I'll mark this as resolved.